---
title: "Assignment 4 BINF 6210: Analysing differences in gene expression related to pre-menstrual dysphoric disorder (PMDD) using a mouse model"
output: By Eva Burguete-Innocente
date: December 3rd, 2025
---

###### Part 1: Analysing the data as per Law et al. (2018) Bioconductor vignette ######

This dataset is from Marrocco et al. (2018) and is available on NCBI GEO under the accession GSE121412.
It is the raw count values of gene expression levels generated by whole-genome RNA-sequencing of the mouse ventral hippocampus. RNA was extracted by decapitation after six weeks of treatment with either vehicle (i.e., no treatment) or estradiol, a hormone. Some mice are wild type and some are heterozygous at position 66 of the brain-derived neurotrophic factor gene, BDNF. This SNP is called Val66Met as it results in an amino acid change from valine to methionine.
The data contains ten samples: four WT mice (of which two received no treatment (vehicle), and two received estradiol (E2)), and six Het-Met mice (of which three received vehicle and three received E2).
Gene expression analysis was performed at the exon level.
There are multiple exons for each transcript for each gene in the dataset.
Sample names in the dataset begin with either VV (WT) or VM (Het-Met) followed by an underscore, then Veh (treated with vehicle) or E2 (estradiol), followed by a number (or period followed by number) from 1-3 (the number of samples that belong to each level).

(Marrocco et al., 2018)

Link to my github: https://github.com/evainnocente/PMDD_RNA-seq_analyses.git

#### Step 01: Setup ####

Loading the required packages recommended by the vignette, using R version 4.4.2.

```{r}
# BiocManager::install("limma")
library(limma)

# BiocManager::install("Glimma")
library(Glimma)

# BiocManager::install("edgeR")
library(edgeR)

# install.packages("RColorBrewer")
library(RColorBrewer) # for plot colours

library(tidyverse)

library(readxl) # for reading in the data in excel file format

options(scipen = 99) # I run this so that the counts aren't converted to scientific notation which could result in negative values
```

#### Step 02: Reading in the raw count data ####

I read in the raw count data that I downloaded from accession GSE121412 on November 21st, 2025 and filtered it to remove NAs and combine the Gene ID, Transcript ID, and Exon ID columns into a unique identifier. I converted the data into a DGE list object for analyses using the edgeR package. 

```{r}
df_raw <- read_excel("../data/GSE121412_OVX_BDNF_E2_All_Raw_Values.xlsx")

# Checking what the data looks like
head(df_raw)
dim(df_raw)
class(df_raw)

# Convert to dataframe
df_raw <- as.data.frame(df_raw)

# Check for NAs in the columns

df_raw %>%
  filter(is.na(`Gene ID`)) %>%
  count()

df_raw %>%
  filter(`Transcript ID` == "NA") %>%
  count()

df_raw %>%
  filter(`Exon ID` == "NA") %>%
  count()

# There are NAs- need to filter dataset

df_raw_matrix <- df_raw %>%
  rename_with(~ gsub(" ", "_", .x)) %>% # Remove spaces in colnames
  filter(!is.na(Gene_ID)) %>% # Remove NAs
  filter(!Transcript_ID == "NA") %>% # NAs in these columns Ire coded as characters, so I put NA in quotes
  filter(!Exon_ID == "NA") %>%
  dplyr::select(1:3, 5:14) %>% # Select relevant columns only
  unite(unique_ID, Gene_ID, Transcript_ID, Exon_ID, sep = "_", remove = T) %>% # Create new column combining gene, transcript, and exon information separated
  column_to_rownames("unique_ID") # Put this information as row names for the matrix

# Convert to matrix
df_raw_matrix <- as.matrix(df_raw_matrix)

# Convert to DGE list object in the package edgeR
df_raw_dge <- DGEList(counts = df_raw_matrix)
```


#### Step 03: Reading in sample metadata ####

VV/WT = Wild Type, VM/Het-Met = Heterozygous BDNF Val66Met , E2 = Estradiol, Veh = Vehicle, OVX = Ovariectomized (all mice Ire ovariectomised)

I captured metadata from the series matrix file by hand from GEO and saved it as a CSV. I included the sample names, accession numbers, genotype information, and treatment information. I read in the .csv and inserted the columns into the DGE list object. I also created vectors of names in these columns for future plotting and naming purposes.


```{r}
df_sample <- read_csv("../data/sample_info.csv")

# Renamed the columns of the DGE list object to the sample names
colnames(df_raw_dge) <- df_sample$sample_name

# Group is a vector of the genotype and treatment together- this is necessary for the contrast matrix later
group <- as.factor(c("WT_Vehicle", "WT_Vehicle", "WT_Estradiol", "WT_Estradiol", "BDNF_Met_Vehicle", "BDNF_Met_Vehicle", "BDNF_Met_Vehicle", "BDNF_Met_Estradiol", "BDNF_Met_Estradiol", "BDNF_Met_Estradiol"))

# Genotype
genotype <- as.factor(c("WT", "WT", "WT", "WT", "BDNF_Met", "BDNF_Met", "BDNF_Met", "BDNF_Met", "BDNF_Met", "BDNF_Met"))

# Treatment
treatment <- as.factor(c("Vehicle", "Vehicle", "Estradiol", "Estradiol", "Vehicle", "Vehicle", "Vehicle", "Estradiol", "Estradiol", "Estradiol"))

# Inserting the information
df_raw_dge$samples$group <- group
df_raw_dge$samples$treatment <- df_sample$treatment
df_raw_dge$samples$genotype <- df_sample$genotype
```


While Law et al. (2018) add gene annotation information to their data, I could not do this because my data is quantified at the exon level. So, I skip adding annotation for now. 

#### Step 04: Log transforming the data ####

Log transformation is necessary to account for differences in counts due to some libraries being sequenced at greater depths (Law et al., 2018). 

```{r}
lcpm <- cpm(df_raw_dge, log = TRUE)

L <- mean(df_raw_dge$samples$lib.size) * 1e-6 # Check the mean and median library size
M <- median(df_raw_dge$samples$lib.size) * 1e-6
c(L, M)

summary(lcpm)
```

#### Step 05: Remove genes that are lowly expressed ####

I want to remove genes with counts of 0 because they are not of interest to me, and they may cause statistical inaccuracies further downstream (Law et al., 2018). I also want to remove genes that have very low levels of expression.

```{r}
table(rowSums(df_raw_dge$counts == 0) == 9) # Check how many exons have an expression level of 0
# 4702 have an expression level of 0

keep.exprs <- filterByExpr(df_raw_dge, group = group) # Filter out exons with less than 10 read counts in each unique combination of samples
df_raw_dgefilt <- df_raw_dge[keep.exprs, , keep.lib.sizes = FALSE] # Here I apply the filter to rows, retain all columns, and recalculate library sizes based on the new number of reads per sample

dim(df_raw_dgefilt) # Now contains 170800 exons
dim(df_raw_dge) # Before filtering, I had 389571
```

#### Step 06: Plot the distribution of log-CPM values ####

These plots show the density of log-CPM values for all samples before and after filtering. The dashed line is the filtering threshold value. The curve of the plot becomes flatter after filtering out 0 and low count exons.

```{r}
samplenames <- colnames(df_raw_dge)

lcpm.cutoff <- log2(10 / M + 2 / L) # This is a similar calculation to the filterbyExpr function did, just for plotting purposes

# Plotting unfiltered data (i.e. 0 and low counts retained)
nsamples <- ncol(df_raw_dge) # colours each sample differently
col <- brewer.pal(nsamples, "Set2") # Specify palette for plotting

par(mfrow = c(1, 2))
plot(density(lcpm[, 1]), col = col[1], lwd = 2, ylim = c(0, 0.26), las = 2, main = "", xlab = "")
title(main = "A. Raw data", xlab = "Log-cpm")
abline(v = lcpm.cutoff, lty = 3)
for (i in 2:nsamples) {
  den <- density(lcpm[, i])
  lines(den$x, den$y, col = col[i], lwd = 2)
}
legend("topright", samplenames, text.col = col, bty = "n")

# Re-run log-transformation with filtered data
lcpmfilt <- cpm(df_raw_dgefilt, log = TRUE)

# Plot filtered data
plot(density(lcpmfilt[, 1]), col = col[1], lwd = 2, ylim = c(0, 0.26), las = 2, main = "", xlab = "")
title(main = "B. Filtered data", xlab = "Log-cpm")
abline(v = lcpm.cutoff, lty = 3)
for (i in 2:nsamples) {
  den <- density(lcpmfilt[, i])
  lines(den$x, den$y, col = col[i], lwd = 2)
}
legend("topright", samplenames, text.col = col, bty = "n") # The plot does not spike up as much after filtering
```

#### Step 07: Normalising the distribution of gene expression ####

This accounts for external factors influencing the expression counts in different samples. Normalisation makes sure that all samples have similar distributions of expression counts. The trimmed mean of M-values (TMM) is used. It is the weighted trimmed mean of the log-ratios of genes between two samples (Robinson and Oshlack, 2010).

```{r}
df_raw_dgenorm <- calcNormFactors(df_raw_dgefilt, method = "TMM")

df_raw_dgenorm$samples$norm.factors # View the normalisation factors, they did not change much from 1
```

#### Step 08: Plotting the difference between normalised and un-normalised data ####

There is not much visual difference between the normalised and un-normalised data, indicating that TMM did not change the data much. However, it is still important to normalise.

```{r}
# Plot un-normalised
par(mfrow = c(1, 2))
lcpm <- cpm(df_raw_dgefilt, log = TRUE)
un_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "A. Unnormalised data", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.5, labels = samplenames, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)


# Plot normalised
lcpm <- cpm(df_raw_dgenorm, log = TRUE)
tmm_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "B. TMM normalised data", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.5, labels = samplenames, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)
```

#### Step 09: Unsupervised clustering of samples ####

These are multi-dimensional scaling plots used to determine how samples cluster relative to their groups (treatment or genotype). If samples are more clustered, they will likely show more differential expression (Law et al., 2018).

```{r}
# Output figure
 png("../figs/unsupervised_clustering_TMMplot.png", width = 1400, height = 800, res = 150)

par(mfrow = c(1, 2), cex.main = 1)

# First plot genotype over dimension 1 and 2
lcpm <- cpm(df_raw_dgenorm, log = TRUE)
col.group <- genotype
levels(col.group) <- c("red", "blue")
col.group <- as.character(col.group)
plotMDS(lcpm, labels = genotype, col = col.group) # Default plots over dim 1 and 2, dim 1 explains 16% of leading log-fold change and dim 2 13%, samples cluster more
# plotMDS(lcpm, labels=genotype, col=col.group, dim = c(3,4)) # Didn't cluster better than dim 1 and 2, each dimension only explained 11% of the log-fold change
title(main = "A. Genotypes")

# Plot treatment over dimension 3 and 4
col.treatment <- treatment
levels(col.treatment) <- c("yellow3", "purple")
col.treatment <- as.character(col.treatment)
# plotMDS(lcpm, labels=treatment, col=col.treatment) # Doesn't cluster much
plotMDS(lcpm, labels = treatment, col = col.treatment, dim = c(3, 4)) # Slightly better clustering
title(main = "B. Treatments")

 dev.off()
```

#### Step 10: Differential expression analysis- creating a design matrix and contrasts ####

I followed Law et al. (2020) to construct the design matrix and set up contrasts.

The columns of the matrix represent:
1) overall differences between treatments, E2 vs. Vehicle 
2) overall differences between genotypes, WT versus BDNF_Met; 
3) differences between treatments within WT genotype, 
4) differences between treatments within BDNF_Met
5) differences between genotypes within Vehicle
6) differences between genotypes within E2. 

```{r}
design <- model.matrix(~ 0 + group) # Setting up the design matrix excluding the intercept term
design # Check how it looks

contrasts <- makeContrasts(
  VehvsE2 = (groupWT_Vehicle + groupBDNF_Met_Vehicle) / 2 - (groupWT_Estradiol + groupBDNF_Met_Estradiol) / 2, # Have to divide by 2 because we are comparing across all samples and need the average
  WTvsBDNF = (groupWT_Vehicle + groupWT_Estradiol) / 2 - (groupBDNF_Met_Vehicle + groupBDNF_Met_Estradiol) / 2,
  VehvsE2_WT = groupWT_Vehicle - groupWT_Estradiol,
  VehvsE2_BDNF_Met = groupBDNF_Met_Vehicle - groupBDNF_Met_Estradiol,
  WTvsBDNF_Met_Veh = groupWT_Vehicle - groupBDNF_Met_Vehicle,
  WTvsBDNF_Met_E2 = groupWT_Estradiol - groupBDNF_Met_Estradiol,
  levels = colnames(design)
) # Specify column names
rownames(contrasts) <- gsub("group", "", rownames(contrasts)) # Remove word group from row names for clarity
contrasts # Check how it looks
```

#### Step 11: Removing heteroscedascity from count data ####

Use voom to account for the non-independent mean-variance relationship of RNA-seq data (Law et al., 2018, Law et al., 2014)

```{r}
par(mfrow = c(1, 2))
v <- voom(df_raw_dgenorm, design, plot = TRUE) # Taking the LCPM with precision weights and plots it
v

vfit <- lmFit(v, design) # Fits a linear model for each exon
vfit <- contrasts.fit(vfit, contrasts = contrasts) # Applies model to contrasts
efit <- eBayes(vfit) # Computes t-values that rank the genes for more precise estimates

# Plots the mean-variance relationship explained by the model
plotSA(efit, main = "Final model: Mean-variance trend")
```

#### Step 12: Calculating how many differentially expressed genes there are ###

The decideTests function classifies the test statistics from the previous models as significantly up- or down- regulated while accounting for multiple testing with the Benjamini-Hochberg correction. I made Venn diagrams to see the number of significantlt DE genes between the contrasts that has significant genes.

```{r}
de_genes <- decideTests(efit) # Shows which exons are up- or down-regulated for each contrast
summary(de_genes) # There are only 2 significantly down-regulated exons between WT vs. BDNF_Met overall, and 3 down-regulated and 2 upregulated for WT vs. BDNF_Met with Vehicle treatment

de.common <- which(de_genes[, 1:6] != 0) # Total exons in DE
length(de.common) # 7

vennDiagram(de_genes[, c(2, 5)], circle.col = c("blue", "red")) # One exon is commonly in DE between these two

# write.fit(efit, de_genes, file="../outputs/de_results_TMM.txt") # Write to output
```


#### Step 13: Looking at exons in DE ####

The topTable function shows the top-ranked exons from the linear model. In the model summary, the only contrasts with significant DE exons were columns 2 and 5, so those are specified with the coef paramater. I ranked the results by adjusted p-value, using a cutoff of 0.05. The p-value is adjusted with the BH correction by default. 
For WT vs BDNF_Met, the two DE exons are 22375_uc007pai.2_1 and 22375_uc007pai.2_3. For WT vs. BDNF_Met for Vehicle, the five DE exons are 64930_uc008iyx.1_23, 207728_uc009iot.3_28, 22375_uc007pai.2_1, 64930_uc008iyz.1_23, 330119_uc012dyf.1_1. 

```{r}
WT_vs_BDNF <- topTable(efit, coef = 2, n = 10, sort.by = "p", p.value = 0.05)
WTvsBDNF_Met_Veh <- topTable(efit, coef = 5, n = 10, sort.by = "p", p.value = 0.05)
WT_vs_BDNF # Display results
WTvsBDNF_Met_Veh

# Writing results

WTvsBDNF <- WT_vs_BDNF %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF", .before = 1)

WTvsBDNF_Veh <- WTvsBDNF_Met_Veh %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF_Met_Veh", .before = 1)

DE_lcpm_TMM <- bind_rows(WTvsBDNF, WTvsBDNF_Veh)

# write.table(DE_lcpm_TMM, "../outputs/DE_TMM_exon_results.txt")
```


#### Step 14: Interactive plots of the DE results ####

I used glimma to make interactive plots of the DE results for each contrast. Differentially expressed genes are shown in red (up) or blue (down)

```{r}
# Change launch = F to T in order to launch plot in browser

# Shows exons in DE for the WT vs. BDNF_Met genotype contrast
WTvsBDNF_Met <- glMDPlot(efit, coef = 2, status = de_genes, main = colnames(efit)[2], counts = lcpm, groups = samplenames, fontsize = 8, html = "WTvsBDNF_Met_DE_TMMplot", launch = F)

# Shows exons in DE for the WT vs. BDNF_Met genotype contrast with Vehicle treatment
WTvsBDNF_Met_Vehicle <- glMDPlot(efit, coef = 5, status = de_genes, html = "WTvsBDNF_Met_Veh_DE_TMMplot", main = colnames(efit)[5], counts = lcpm, groups = samplenames, launch = F)

# The plots are automatically written to a folder called glimma-plots in the working directory, and the location cannot be changed. Although I moved the folder myself to the figs folder after I finished running the analyses, when the script is rerun by someone else the glimma-plots folder will regenerate in the working directory (R folder) (just a heads up). 
```

###### Part 2: Exploring the consequences of varying normalisation method ######

I will change the method of normalisation from TMM to upperquartile, and see what effects it has on the output. I will start the analyses from Step 07, as all steps before that remain the same.

#### Step 07.2: Normalising the distribution of gene expression ####

I am going to change the normalisation method here from TMM to upper quartile. Scale factors are calculated from the 75% quantile of each library counts, and genes that are 0 are automatically removed (Bullard et al., 2010).

```{r}
df_raw_dgenorm_uq <- calcNormFactors(df_raw_dgefilt, method = "upperquartile")

df_raw_dgenorm_uq$samples$norm.factors # View the normalisation factors, they did not change much from 1
```

#### Step 08.2: Plotting the difference between normalised and un-normalised data ####

As before, there is not much visual difference between the normalised and un-normalised data, indicating that upper quartile did not change the data much (that we can see).

```{r}
# Plot un-normalised
par(mfrow = c(1, 2))
lcpm <- cpm(df_raw_dgefilt, log = TRUE)
un_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "A. Unnormalised data", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.5, labels = group, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)

# Plot normalised
lcpm <- cpm(df_raw_dgenorm_uq, log = TRUE)
uq_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "B.Upper quartile normalised data", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.5, labels = group, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)

# For comparison of all three, a figure of TMM normalisation, UQ normalisation, and no normalisation:

# To output the multi-panel figure
# png("../figs/multi_panel_normplot.png", width = 1600, height = 1200, res = 150)

par(mfrow = c(1, 3), mar = c(5, 4, 2, 2))

lcpm <- cpm(df_raw_dgefilt, log = TRUE)
un_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "A. No normalisation", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.3, labels = samplenames, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)

lcpm <- cpm(df_raw_dgenorm, log = TRUE)
tmm_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "B. TMM normalisation", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.3, labels = samplenames, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)

lcpm <- cpm(df_raw_dgenorm_uq, log = TRUE)
uq_norm <- boxplot(lcpm, las = 2, col = col, main = "", ylim = c(-5, 15), names = F)
title(main = "C. Upper quartile normalisation", ylab = "Log-cpm")
text(x = 1:length(group), par("usr")[3] - 0.3, labels = samplenames, cex = 0.7, srt = 45, adj = c(1, 1), xpd = TRUE)

# dev.off()
```

#### Step 09.2: Unsupervised clustering of samples ####

As before, I made MDS plots to show how samples cluster. The samples cluster slightly differently than TMM.

```{r}
# First plot genotype over dimension 1 and 2
lcpm <- cpm(df_raw_dgenorm_uq, log = TRUE)
par(mfrow = c(1, 2))
col.group <- genotype
levels(col.group) <- c("red", "blue")
col.group <- as.character(col.group)
plotMDS(lcpm, labels = genotype, col = col.group) # Default plots over dim 1 and 2, doesn't cluster much
# plotMDS(lcpm, labels=genotype, col=col.group, dim = c(3,4)) # Didn't cluster better than dim 1 and 2
title(main = "A. Genotypes")

# Plot treatment over dimension 3 and 4
col.treatment <- treatment
levels(col.treatment) <- c("red", "blue")
col.treatment <- as.character(col.treatment)
# plotMDS(lcpm, labels=treatment, col=col.treatment) # Doesn't cluster much
plotMDS(lcpm, labels = treatment, col = col.treatment, dim = c(3, 4)) # Slightly better clustering
title(main = "B. Treatments")
```

#### Step 10.2: Differential expression analysis- creating a design matrix and contrasts ####

Constructing design matrix and contrasts as before.

```{r}
design <- model.matrix(~ 0 + group) # Setting up the design matrix excluding the intercept term
design # Check how it looks

contrasts <- makeContrasts(
  VehvsE2 = (groupWT_Vehicle + groupBDNF_Met_Vehicle) / 2 - (groupWT_Estradiol + groupBDNF_Met_Estradiol) / 2, # Have to divide by 2 because we are comparing across all samples and need the average
  WTvsBDNF = (groupWT_Vehicle + groupWT_Estradiol) / 2 - (groupBDNF_Met_Vehicle + groupBDNF_Met_Estradiol) / 2,
  VehvsE2_WT = groupWT_Vehicle - groupWT_Estradiol,
  VehvsE2_BDNF_Met = groupBDNF_Met_Vehicle - groupBDNF_Met_Estradiol,
  WTvsBDNF_Met_Veh = groupWT_Vehicle - groupBDNF_Met_Vehicle,
  WTvsBDNF_Met_E2 = groupWT_Estradiol - groupBDNF_Met_Estradiol,
  levels = colnames(design)
) # Specify column names
rownames(contrasts) <- gsub("group", "", rownames(contrasts)) # Remove word group from row names for clarity
contrasts # Check how it looks
```

#### Step 11.2: Removing heteroscedascity from count data ####

Creating voom plot and models as before.

```{r}
par(mfrow = c(1, 2))
v <- voom(df_raw_dgenorm_uq, design, plot = TRUE) # Taking the LCPM with precision weights and plots it
v

vfit <- lmFit(v, design) # Fits a linear model for each exon
vfit <- contrasts.fit(vfit, contrasts = contrasts) # Applies model to contrasts
efit <- eBayes(vfit) # Computes t-values that rank the genes for more precise estimates

# Plots the mean-variance relationship explained by the model
plotSA(efit, main = "Final model: Mean-variance trend")
```

#### Step 12.2: Calculating how many differentially expressed exons there are ###

Evaluating numbers of DE exons as before. The same numbers of exons in the same contrasts are in DE, but further examination is needed to see whether they are the same exons as before.

```{r}
de_genes <- decideTests(efit) # Shows which exons are up- or down-regulated for each contrast
summary(de_genes) # There are only 2 significantly down-regulated exons between WT vs. BDNF_Met overall, and 3 down-regulated and 2 upregulated for WT vs. BDNF_Met with Vehicle treatment

de.common <- which(de_genes[, 1:6] != 0) # Total exons in DE
length(de.common) # 7

vennDiagram(de_genes[, c(2, 5)], circle.col = c("blue", "red")) # One exon is commonly in DE between these two

# write.fit(efit, de_genes, file="../outputs/de_results_UQ.txt") # Write to output
```

#### Step 13.2: Looking at exons in DE ####

Displaying top ranked DE exons for each contrast as before
For WT vs BDNF_Met, the two DE exons are 22375_uc007pai.2_1 and 22375_uc007pai.2_3, the same as before. For WT vs. BDNF_Met for Vehicle, the five DE exons are the same as before, but interestingly, the ranking switched around slightly. The top ranked exons are now in this order: 64930_uc008iyx.1_23, 207728_uc009iot.3_28, 64930_uc008iyz.1_23, 22375_uc007pai.2_1, 330119_uc012dyf.1_1. 

```{r}
WT_vs_BDNF <- topTable(efit, coef = 2, n = 10, sort.by = "p", p.value = 0.05)
WTvsBDNF_Met_Veh <- topTable(efit, coef = 5, n = 10, sort.by = "p", p.value = 0.05)
WT_vs_BDNF # Display results
WTvsBDNF_Met_Veh

# Writing results

WTvsBDNF <- WT_vs_BDNF %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF", .before = 1)

WTvsBDNF_Veh <- WTvsBDNF_Met_Veh %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF_Met_Veh", .before = 1)

DE_lcpm_UQ <- bind_rows(WTvsBDNF, WTvsBDNF_Veh)

# write.table(DE_lcpm_UQ, "../outputs/DE_UQ_exon_results.txt")
```

#### Step 14.2: Interactive plots of the DE results ####

Again, I used glimma to make interactive plots of the DE results for each contrast. Differentially expressed genes are shown in red (up) or blue (down).

```{r}
# Change launch = F to T in order to launch plot in browser

WTvsBDNF_Met <- glMDPlot(efit, coef = 2, status = de_genes, main = colnames(efit)[2], counts = lcpm, groups = samplenames, fontsize = 4, html = "WTvsBDNF_Met_DE_UQplot", launch = F)

# Shows exons in DE for the WT vs. BDNF_Met genotype contrast with Vehicle treatment
WTvsBDNF_Met_Vehicle <- glMDPlot(efit, coef = 5, status = de_genes, html = "WTvsBDNF_Met_Veh_DE_UQplot", main = colnames(efit)[5], counts = lcpm, groups = samplenames, launch = F)
```

###### Part 3: Exploring the consequences of not normalising data at all ######

I will change the method of normalisation to none, to see what effects it has on the output. Again, I will start the analyses from Step 07, as all steps before that remain the same.

#### Step 07.3: (Not) Normalising the distribution of gene expression ####

I specify the normalisation method as "none". This means that the distribution of expression counts for each sample is not being adjusted at all, and external factors could be influencing the data. This time I am skipping plotting the un-normalised data, as we have already seen that in Parts 1 and 2 above.

```{r}
df_raw_dgenorm_none <- calcNormFactors(df_raw_dgefilt, method = "none")

df_raw_dgenorm_none$samples$norm.factors # View the normalisation factors, they are all 1 because no scaling factor is being applied
```

#### Step 08.3: Unsupervised clustering of samples ####

As before, I made MDS plots to show how samples cluster. The samples cluster slightly differently than both prior normalisation methods.

```{r}
# First plot genotype over dimension 1 and 2
lcpm <- cpm(df_raw_dgenorm_none, log = TRUE)
par(mfrow = c(1, 2))
col.group <- genotype
levels(col.group) <- c("red", "blue")
col.group <- as.character(col.group)
plotMDS(lcpm, labels = genotype, col = col.group) # Default plots over dim 1 and 2, doesn't cluster much
# plotMDS(lcpm, labels=genotype, col=col.group, dim = c(3,4)) # Didn't cluster better than dim 1 and 2
title(main = "A. Genotypes")

# Plot treatment over dimension 3 and 4
col.treatment <- treatment
levels(col.treatment) <- c("red", "blue")
col.treatment <- as.character(col.treatment)
# plotMDS(lcpm, labels=treatment, col=col.treatment) # Doesn't cluster much
plotMDS(lcpm, labels = treatment, col = col.treatment, dim = c(3, 4)) # Slightly better clustering
title(main = "B. Treatments")
```

#### Step 09.3: Differential expression analysis- creating a design matrix and contrasts ####

Again, constructing design matrix and contrasts as before.

```{r}
design <- model.matrix(~ 0 + group) # Setting up the design matrix excluding the intercept term
design # Check how it looks

contrasts <- makeContrasts(
  VehvsE2 = (groupWT_Vehicle + groupBDNF_Met_Vehicle) / 2 - (groupWT_Estradiol + groupBDNF_Met_Estradiol) / 2, # Have to divide by 2 because we are comparing across all samples and need the average
  WTvsBDNF = (groupWT_Vehicle + groupWT_Estradiol) / 2 - (groupBDNF_Met_Vehicle + groupBDNF_Met_Estradiol) / 2,
  VehvsE2_WT = groupWT_Vehicle - groupWT_Estradiol,
  VehvsE2_BDNF_Met = groupBDNF_Met_Vehicle - groupBDNF_Met_Estradiol,
  WTvsBDNF_Met_Veh = groupWT_Vehicle - groupBDNF_Met_Vehicle,
  WTvsBDNF_Met_E2 = groupWT_Estradiol - groupBDNF_Met_Estradiol,
  levels = colnames(design)
) # Specify column names
rownames(contrasts) <- gsub("group", "", rownames(contrasts)) # Remove word group from row names for clarity
contrasts # Check how it looks
```

#### Step 10.3: Removing heteroscedascity from count data ####

Again, creating voom plot and models as before.

```{r}
par(mfrow = c(1, 2))
v <- voom(df_raw_dgenorm_none, design, plot = TRUE) # Taking the LCPM with precision weights and plots it
v

vfit <- lmFit(v, design) # Fits a linear model for each exon
vfit <- contrasts.fit(vfit, contrasts = contrasts) # Applies model to contrasts
efit <- eBayes(vfit) # Computes t-values that rank the genes for more precise estimates

# Plots the mean-variance relationship explained by the model
plotSA(efit, main = "Final model: Mean-variance trend")
```

#### Step 11.3: Calculating how many differentially expressed exons there are ####

Evaluating numbers of DE exons as before. This time, the results are different. For the WTvsBDNF contrast, 3 exons are significantly down-regulated, instead of 2. For the WTvsBDNF_Met_Veh contrast, 1 gene was significantly down-regulated and 1 was significantly up-regulated, whereas before, 3 were down and 2 were up. As well, there are now genes in DE for the VehvsE2_WT contrast: 1 up- and 1 down-regulated. However, the overall number of exons in DE remains the same. 

```{r}
de_genes <- decideTests(efit) # Shows which exons are up- or down-regulated for each contrast
summary(de_genes) # Different from before

de.common <- which(de_genes[, 1:6] != 0) # Total exons in DE
length(de.common) # Still 7 overall

vennDiagram(de_genes[, c(2, 3, 5)], circle.col = c("blue", "red")) # One exon is commonly in DE in all three, and 1 between VehvsE2_WT and WTvsBDNF_Met_Veh, as before

# write.fit(efit, de_genes, file="../outputs/de_results_none.txt") # Write to output
```

#### Step 12.3: Looking at exons in DE ####

Displaying top ranked DE exons for each contrast as before
For WT vs BDNF_Met, the two DE exons are 22375_uc007pai.2_1 and 22375_uc007pai.2_3, the same as before. For WT vs. BDNF_Met for Vehicle, the five DE exons are the same as before, but interestingly, the ranking switched around slightly. The top ranked exons are now in this order: 64930_uc008iyx.1_23, 207728_uc009iot.3_28, 64930_uc008iyz.1_23, 22375_uc007pai.2_1, 330119_uc012dyf.1_1. 

```{r}
WT_vs_BDNF <- topTable(efit, coef = 2, n = 10, sort.by = "p", p.value = 0.05)
VehvsE2_WT <- topTable(efit, coef = 3, n = 10, sort.by = "p", p.value = 0.05)
WTvsBDNF_Met_Veh <- topTable(efit, coef = 5, n = 10, sort.by = "p", p.value = 0.05)
WT_vs_BDNF # Display results
VehvsE2_WT
WTvsBDNF_Met_Veh

# Writing results

WTvsBDNF <- WT_vs_BDNF %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF", .before = 1)

VehvsE2WT <- VehvsE2_WT %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "VehvsE2_WT", .before = 1)

WTvsBDNF_Veh <- WTvsBDNF_Met_Veh %>%
  rownames_to_column(var = "exon_name") %>%
  mutate(contrast = "WTvsBDNF_Met_Veh", .before = 1)

DE_lcpm_none <- bind_rows(WTvsBDNF, VehvsE2WT, WTvsBDNF_Veh)

# write.table(DE_lcpm_none, "../outputs/DE_none_exon_results.txt")
```

#### Step 13.3: Interactive plots of the DE results ####

Again, I used glimma to make interactive plots of the DE results for each contrast. Differentially expressed genes are shown in red (up) or blue (down)

```{r}
# Change launch = F to T in order to launch plot in browser

glMDPlot(efit, coef = 2, status = de_genes, html = "WTvsBDNF_Met_DE_none_plot", main = colnames(efit)[2], counts = lcpm, groups = samplenames, launch = F) # Shows exons in DE for the WT vs. BDNF_Met genotype contrast

glMDPlot(efit, coef = 3, status = de_genes, html = "VehvsE2_WT_DE_none_plot", main = colnames(efit)[3], counts = lcpm, groups = samplenames, launch = F) # Shows exons in DE for the Veh vs E2 treatment in WT genotype contrast

glMDPlot(efit, coef = 5, status = de_genes, html = "WTvsBDNF_Met_Veh_DE_none_plot", main = colnames(efit)[5], counts = lcpm, groups = samplenames, launch = F) # Shows exons in DE for the WT vs. BDNF_Met genotype contrast with Vehicle treatment
```

```{r}
# Session information
sessionInfo()
```

#### References ####
Bullard, J. H., Purdom, E., Hansen, K. D., & Dudoit, S. (2010). Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments. BMC Bioinformatics, 11(1), 94. https://doi.org/10.1186/1471-2105-11-94

Law, C. W., Y. Chen, W. Shi, and G. K. Smyth. 2014. “Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-seq Read Counts.” Genome Biology 15: R29.

Law, C. W., Alhamdoosh, M., Su, S., Dong, X., Tian, L., Smyth, G. K., & Ritchie, M. E. (2018). RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR. https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html

Law, C., Zeglinski, K., Dong, X., Alhamdoosh, M., Smyth, G. K., & Ritchie, M. E. (2020). A guide to creating design matrices for gene expression experiments. https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html

Marrocco, J., Einhorn, N. R., Petty, G. H., Li, H., Dubey, N., Hoffman, J., Berman, K. F., Goldman, D., Lee, F. S., Schmidt, P. J., & McEwen, B. S. (2020). Epigenetic intersection of BDNF Val66Met genotype with premenstrual dysphoric disorder transcriptome in a cross-species model of estradiol add-back. Molecular Psychiatry, 25(3), 572–583. https://doi.org/10.1038/s41380-018-0274-3

Ritchie, M. E., Phipson, B., Wu, D., Hu, Y., Law, C. W., Shi, W., & Smyth, G. K. (2015). Limma powers differential expression analyses for rna-sequencing and microarray studies. Nucleic Acids Research, 43(7), e47. https://doi.org/10.1093/nar/gkv007

Robinson, M. D., McCarthy, D. J., & Smyth, G. K. (2010). Edger: A bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26(1), 139–140. https://doi.org/10.1093/bioinformatics/btp616

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3), R25. https://doi.org/10.1186/gb-2010-11-3-r25

Su, S., Law, C. W., Ah-Cann, C., Asselin-Labat, M.-L., Blewitt, M. E., & Ritchie, M. E. (2017). Glimma: Interactive graphics for gene expression analysis. Bioinformatics (Oxford, England), 33(13), 2050–2052. https://doi.org/10.1093/bioinformatics/btx094

Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, Grolemund G, Hayes A, Henry L, Hester J, Kuhn M, Pedersen TL, Miller E, Bache SM, Müller K, Ooms J, Robinson D, Seidel DP, Spinu V, Takahashi K, Vaughan D, Wilke C, Woo K, Yutani H (2019). “Welcome to the tidyverse.” Journal of Open Source Software, 4(43), 1686. doi:10.21105/joss.01686.
